pkg/operator/metrics/collector.go

```diff
@@ -0,0 +1,215 @@
+	workkey                     = "key"
+)
+
+// OperatorStateGetter knows how to get the current operator state and provides operator client informer.
+type OperatorStateGetter interface {
+	GetOperatorState() (spec *operatorv1.OperatorSpec, status *operatorv1.OperatorStatus, staticPodStatus *operatorv1.StaticPodOperatorStatus, resourceVersion string, err error)
+	Informer() cache.SharedIndexInformer
+}
+
+// operatorMetricUpdater knows how to update a single tracked metric.
+type operatorMetricUpdater interface {
+	Update(*operatorv1.OperatorSpec, *operatorv1.OperatorStatus, *operatorv1.StaticPodOperatorStatus) error
+}
+
+type operatorMetricUpdateFn func(*operatorv1.OperatorSpec, *operatorv1.OperatorStatus, *operatorv1.StaticPodOperatorStatus) ([]metricValue, error)
+
+type Collector struct {
+	operatorClient  OperatorStateGetter
+	operatorMetrics []operatorMetricUpdater
+	targetNamespace string
+
+	queue          workqueue.RateLimitingInterface
+	hasSynced      cache.InformerSynced
+	collectorMutex sync.RWMutex
+}
+
+// metricValue represents a single metric value sample
+type metricValue struct {
+	value  float64
+	labels []string
+}
+
+// operatorMetrics represents a single metric definition
+type operatorMetric struct {
+	prometheusDesc *prometheus.Desc
+	valueType      prometheus.ValueType
+
+	// values are the recently observed metric values provided by the updateFn
+	values []metricValue
+
+	// updateFn know how to gather the values for this metric
+	updateFn operatorMetricUpdateFn
+
+	// valueMutex protects from values changes
+	valueMutex sync.RWMutex
+}
+
+// Update updates the values for the current metric.
+func (m *operatorMetric) Update(spec *operatorv1.OperatorSpec, status *operatorv1.OperatorStatus, staticPodStatus *operatorv1.StaticPodOperatorStatus) error {
+	m.valueMutex.Lock()
+	defer m.valueMutex.Unlock()
+	currentValues, err := m.updateFn(spec, status, staticPodStatus)
+	if err != nil {
+		return err
+	}
+	m.values = currentValues
+	return nil
+}
+
+// NewOperatorStatusMetricsCollector returns a metrics collector that automatically collect prometheus metrics from given operator status.
+func NewOperatorStatusMetricsCollector(client OperatorStateGetter, targetOperatorNamespace string) *Collector {
+	collector := &Collector{
+		operatorClient:  client,
+		targetNamespace: targetOperatorNamespace,
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "OperatorStatusMetricsCollector"),
+		hasSynced:       client.Informer().HasSynced,
+	}
+
+	// register default set of metrics we track
+	addStatusConditionsMetrics(collector)
+	addStaticPodStatusLatestRevisionMetrics(collector)
+	addStaticPodNodesMetrics(collector)
+
+	client.Informer().AddEventHandler(collector.eventHandler())
+
+	// NOTE: Only call this once and this must be called after all metrics are registered otherwise prometheus will panic.
+	prometheus.MustRegister(collector)
```

Comment:
can and should we wrap this somehow in a `Once` struct? Not being able to call a constructor twice is odd.